src/system.rs

```diff
@@ -138,109 +156,932 @@ impl StageExecutor {
+    unsafe fn prepare(&mut self, _: &World) {}
 }
 
-struct System<
-    R: Accessor,
-    Q: for<'a> QuerySet<'a>,
-    F: for<'a> Fn(R::Output, <Q as QuerySet<'a>>::PreparedQueries),
-> {
-    _resources: PhantomData<R>,
-    queries: Q,
-    run_fn: F,
+impl<AV, AF> QuerySet for Query<AV, AF>
+where
+    AV: for<'v> View<'v>,
+    AF: EntityFilter + Send + Sync,
+{
+    type PreparedQueries = PreparedQuery<AV, AF>;
+    fn filter_archetypes(&mut self, world: &World, bitset: &mut BitSet) {
+        let storage = world.storage();
+        self.filter.iter_archetype_indexes(storage).for_each(|id| {
+            bitset.insert(id);
+        });
+    }
+    unsafe fn prepare(&mut self, world: &World) -> Self::PreparedQueries {
+        PreparedQuery::<AV, AF>::new(world, self)
+    }
+}
+
+impl_queryset_tuple!(A);
+impl_queryset_tuple!(A, B);
+impl_queryset_tuple!(A, B, C);
+impl_queryset_tuple!(A, B, C, D);
+impl_queryset_tuple!(A, B, C, D, E);
+impl_queryset_tuple!(A, B, C, D, E, F);
+impl_queryset_tuple!(A, B, C, D, E, F, G);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y);
+impl_queryset_tuple!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z);
+
+pub struct PreparedWorld {
+    world: *const World,
+    access: *const Access<ComponentTypeId>,
+}
+impl PreparedWorld {
+    unsafe fn new(world: &World, access: &Access<ComponentTypeId>) -> Self {
+        Self {
+            world: world as *const World,
+            access: access as *const Access<ComponentTypeId>,
+        }
+    }
+}
+
+// TODO: these assertions should have better errors
+impl PreparedWorld {
+    #[inline]
+    pub fn get_component<T: Component>(&self, entity: Entity) -> Option<Ref<Shared, T>> {
```

Comment:
This is not going to be sufficient to ensure safety, because the scheduler can schedule two systems that incompatably access the same component type provided that it can determine that they will not access the same set of archetypes.

I had planned on adding these random access APIs to queries, with the queries validating that the access is to an entity that lies within it's own result set.

Alternatively, we could make the reqest for the `PreparedWorld` resource optional and a special case which causes the sheduler to consider queries as accessing all archetypes.

Reply:
So this case specifically needs to take a step back, and consider the use case: This was implemented to be a temporary handrail in the case of optional components. Currently, the query parameter has no concept of `Option` components, e.g. query for a component existing or not-existing. 

This is added for that use case, where I may want to arbitrarily fetch a specific component, and maybe even not associated with a current `Query`. In its current implementation and envisioning, this would be what you classify as a *special case*, in the respect that it should be considered a global access to that component  type.

This is a major consideration in @AThilenius use case for hierarchy transformations. Having optional components depending on the type of transformation and hierarchy, it is much more efficient to "lock" across all archetypes, because a child updating its transform may need to access arbitrary other transforms, and *cannot* be defined by a query.

Reply:
Hi!

I've taken an interest in this crate for use in [a project of mine](https://github.com/caelunshun/feather) (we currently use Specs). I have been keeping an eye on this PR, and I'd like to add that there are other uses for random access for components as well. For example, consider the case where an event is triggered containing an `Entity`, and a system handling that event needs to access components of that entity. Queries currently don't support querying for a single entity or set of entities, so this sort of random access is necessary.

I imagine the solution to that would be to implement a filter for queries which filters out entities not equal to some provided `Entity`.